File: Inheritance/Inheritance.java
/*
 * Runnable examples of inheritance, field hiding, method overriding,
 * constructor order, and access through parent/child references.
 *
 * Compile and run:
 *   javac Inheritance.java
 *   java Inheritance
 */

class Animal {
    int value = 10;
    static String category = "Animal";

    Animal() {
        System.out.println("1. Animal constructor");
    }

    void eat() {
        System.out.println("Animal.eat()");
    }

    void printFields() {
        System.out.println("Animal.value = " + value);
        System.out.println("Animal.category = " + category);
    }
}

class Dog extends Animal {
    // This is a second field. It does not override Animal.value.
    int value = 20;
    static String category = "Dog";

    Dog() {
        super(); // Must be the first statement; initializes the Animal part.
        System.out.println("2. Dog constructor");
    }

    @Override
    void eat() {
        System.out.println("Dog.eat()");
    }

    void bark() {
        System.out.println("Dog.bark()");
    }

    void printFields() {
        System.out.println("Dog.value = " + value);
        System.out.println("Animal.value = " + super.value);
        System.out.println("Dog.category = " + category);
        System.out.println("Animal.category = " + Animal.category);
    }

    int parentValue() {
        return super.value;
    }

    void callParentEat() {
        super.eat();
        eat();
    }
}

public class Inheritance {
    static void separator(String title) {
        System.out.println("\n--- " + title + " ---");
    }

    public static void main(String[] args) {
        separator("Construction order");
        Dog dog = new Dog();

        separator("Access through a child reference");
        System.out.println("dog.value = " + dog.value);             // 20
        System.out.println("((Animal) dog).value = "
                + ((Animal) dog).value);                            // 10
        dog.printFields();                                          // Child method

        separator("Access through a parent reference");
        Animal animal = dog; // Upcasting: safe and implicit
        System.out.println("animal.value = " + animal.value);       // 10
        System.out.println("animal.category = " + animal.category); // Animal
        animal.eat();                                               // Dog.eat()
        animal.printFields();                                       // Dog.printFields()

        separator("Downcasting and parent access");
        if (animal instanceof Dog) {
            Dog sameDog = (Dog) animal;
            sameDog.bark();
            System.out.println("sameDog.value = " + sameDog.value); // 20
            System.out.println("super.value is available only inside Dog: "
                    + sameDog.parentValue());
        }

        separator("Explicit parent method call");
        dog.callParentEat();

        separator("Important rule");
        System.out.println("Fields are hidden, not overridden.");
        System.out.println("Instance methods are overridden and dynamically dispatched.");
        System.out.println("Static fields are hidden and resolved from the reference type.");
    }
}

File: Inheritance/Inheritance.cpp
/*
 * Runnable examples of inheritance, field hiding, method overriding,
 * constructor order, and access through parent/child references.
 *
 * Compile and run:
 *   g++ -std=c++17 -Wall -Wextra -pedantic Inheritance.cpp -o Inheritance
 *   ./Inheritance
 */

#include <iostream>
#include <string>

class Animal {
public:
    int value = 10;
    inline static const std::string category = "Animal";

    Animal() {
        std::cout << "1. Animal constructor\n";
    }

    virtual ~Animal() = default;

    virtual void eat() const {
        std::cout << "Animal::eat()\n";
    }

    void printFields() const {
        std::cout << "Animal::value = " << value << '\n';
        std::cout << "Animal::category = " << category << '\n';
    }
};

class Dog : public Animal {
public:
    // This is a second member. It does not override Animal::value.
    int value = 20;
    inline static const std::string category = "Dog";

    Dog() {
        std::cout << "2. Dog constructor\n";
    }

    void eat() const override {
        std::cout << "Dog::eat()\n";
    }

    void bark() const {
        std::cout << "Dog::bark()\n";
    }

    void printFields() const {
        std::cout << "Dog::value = " << value << '\n';
        std::cout << "Animal::value = " << Animal::value << '\n';
        std::cout << "Dog::category = " << category << '\n';
        std::cout << "Animal::category = " << Animal::category << '\n';
    }

    void callParentEat() const {
        Animal::eat();
        eat();
    }
};

void separator(const char* title) {
    std::cout << "\n--- " << title << " ---\n";
}

int main() {
    separator("Construction order");
    Dog dog;

    separator("Access through a child object");
    std::cout << "dog.value = " << dog.value << '\n';                 // 20
    std::cout << "dog.Animal::value = " << dog.Animal::value << '\n'; // 10
    dog.printFields();                                                 // Dog method

    separator("Access through a parent reference");
    Animal& animal = dog; // Upcasting: safe and implicit
    std::cout << "animal.value = " << animal.value << '\n';            // 10
    std::cout << "animal.category = " << Animal::category << '\n';
    animal.eat();                                                       // Dog::eat()
    animal.printFields();                                               // Animal::printFields()

    separator("Access through a parent pointer");
    Animal* animalPointer = &dog;
    animalPointer->eat();                                               // Dog::eat()
    animalPointer->printFields();                                       // Animal::printFields()
    if (Dog* sameDog = dynamic_cast<Dog*>(animalPointer)) {
        sameDog->bark();
        std::cout << "sameDog->value = " << sameDog->value << '\n';     // 20
    }

    separator("Explicit parent method call");
    dog.callParentEat();

    separator("Important rule");
    std::cout << "Data members are hidden, not dynamically dispatched.\n";
    std::cout << "Virtual methods are overridden and dynamically dispatched.\n";
    std::cout << "Static members are hidden and resolved by the class name.\n";

    return 0;
}
